### PR TITLE
LibRegex: Implement multiline stateful matches

### PR DESCRIPTION
**Pull request by alimpfard**
cc @IdanHo, your test case : )

### CODE REVIEW

File: Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js [hunk 1]
```diff
@@ -155,3 +155,53 @@ test("undefined match result", () => {
...
+`;
+
+    let res = reg.exec(string);
+    expect(res.length).toBe(1);
+    expect(res[0]).toBe("                                // 1:2:3:4:5:6:7::  1:2:3:4:5:6:7:8");
+    expect(res.index).toBe(43);
+});
+
+// This does not work on the second exec:
+// Exception: [TypeError] Cannot get property '0' of null
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js:182]: @linusg: ^

File: Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js [hunk 2]
```diff
@@ -155,3 +155,55 @@ test("undefined match result", () => {
...
+(?:[a-fA-F\d]{1,4}:){3}(?:(:[a-fA-F\d]{1,4}){0,2}:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)){3}|(:[a-fA-F\d]{1,4}){1,4}|:)| // 1:2:3::          1:2:3::5:6:7:8   1:2:3::8        1:2:3::5:6:7:1.2.3.4
+(?:[a-fA-F\d]{1,4}:){2}(?:(:[a-fA-F\d]{1,4}){0,3}:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)){3}|(:[a-fA-F\d]{1,4}){1,5}|:)| // 1:2::            1:2::4:5:6:7:8   1:2::8          1:2::4:5:6:7:1.2.3.4
+(?:[a-fA-F\d]{1,4}:){1}(?:(:[a-fA-F\d]{1,4}){0,4}:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)){3}|(:[a-fA-F\d]{1,4}){1,6}|:)| // 1::              1::3:4:5:6:7:8   1::8            1::3:4:5:6:7:1.2.3.4
+(?::((?::[a-fA-F\d]{1,4}){0,5}:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)){3}|(?::[a-fA-F\d]{1,4}){1,7}|:))           // ::2:3:4:5:6:7:8  ::2:3:4:5:6:7:8  ::8             ::1.2.3.4
+)(%[0-9a-zA-Z]{1,})?                                           // %eth0            %1
```
Comment (linusg) [inline Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js:172]: when people say they don't like regex, this is why :smile:

File: Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js (hunk 1 above)
Comment (alimpfard) [reply Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js:182]: I think this was my fault, I ran it wrong.
exec() just returns null, my bad :sweat_smile:

Comment (linusg) [reply Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js:182]: it does, but should it? when I run this monstrosity in V8, it comes up with a result for both `exec()`s...

Comment (alimpfard) [reply Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js:182]: It shouldn't, I'm trying to figure out why lol

**Comment by IdanHo** [comment]
very nice work :)

**Comment by alimpfard** [comment]
Ended up fixing two (!) extra bugs, yay

**Comment by alimpfard** [comment]
\*sigh\*
Damn you, me, why is Shell-sigpipe failing again...

**Comment by linusg** [comment]
> very nice work :)

I share that sentiment!

**Pull request merged**